### DISCUSSION TITLE
Multiplatform

### DISCUSSION THREAD

**Discussion by elect86** [kotlin-graphics/glm#22] [Ideas]
@ArcheCraft 

So, about matrices, I was wondering if switching to something like

`abstract class Mat2T<V, N> where N : Number, V : Vec2T<out Number>`

This way we might have the component accesses down and having to implement, in the extending classes, only the `abstract operator fun get(index: Int): V` where the implementing matrix class can call the appropriate vector constructor..

**Comment by ArcheCraft** [comment]
Well it shouldn't be `<out Number>` because `Boolean` and the unsigned types don't extend `Number`. That's also something I missed, the `set(index, vector)` shouldn't use `VecXT<out Number>`... But generally that would be possible, except the overhead of creating a Vector and then just reading a value is not necessary as we are anyway just generating the code

**Comment by elect86** [comment]
Yeah right, the point is if we do want to have/makes sense matrices with boolean and unsigned

**Reply by ArcheCraft** [reply]
I created a new branch with only signed matrices, if it is ok we could also merge that or develop on that branch for now, depending of we maybe would like to come back later...

**Reply by ArcheCraft** [reply]
Also I implemented matrix multiplication on that branch

**Reply by ArcheCraft** [reply]
I guess it would be more difficult than that if there were non-signed-number matrices

**Comment by ArcheCraft** [comment]
Yeah, Boolean probably doesn't make sense. I think there are Szenarios in which unsigned matrices would be useful, but I don't know if that has that much use cases that it's worth it. Especially when we could replace all the subclasses with typealiases if we would have only matrices where T extends Number

**Reply by elect86** [reply]
ok, since you've already done it, let's keep it like this unless we come across some scenario where we need something different

**Comment by elect86** [comment]
Also, what about opening up vectors and matrices? One test on the original glm is about extending `vec4`
I'm all for it, but I'd like to hear if you see any drawbacks

**Reply by ArcheCraft** [reply]
I don't see any reason to not do it

**Comment by elect86** [comment]
Also [super interesting](https://youtrack.jetbrains.com/issue/KT-27435) regarding unsigned extending `Number`..

**Reply by ArcheCraft** [reply]
I would guess that still takes some time as target version is 1.7.0 and 1.6.0 basically just came out

**Reply by ArcheCraft** [reply]
Also it only works with interfaces and Number is an abstract class because it is an abstract class on the JVM (Java has it like that) and the expect/actual system doesn't allow changing between classes and interfaces

**Reply by elect86** [reply]
ah yeah, you are right, it's just for interfaces.. :/

**Comment by ArcheCraft** [comment]
Also we should think about if we want to add KDoc to the generated Functions and Classes
I guess that would be helpful to users of the library...

**Reply by elect86** [reply]
Yep, I'm all for it, but I have many more other priorities at the moment (if you want go for it, don't hesitate)
Also we could take inspiration from the [native docs](http://glm.g-truc.net/0.9.8/api/a00146.html#ga693d77696ff36572a0da79efec965acd)

**Comment by ArcheCraft** [comment]
And another thing to think about is if we want to keep the system with a `glm` object and providing the functionality via interfaces as we already started with a different approach (being that we implement the stuff directly on the target class)

**Reply by elect86** [reply]
This should be driven by practical reasons. Let me say that I opted for the interface design to have modular code and give the option to the end user to build its own `glm` handler (similar like on c/c++ where you have specific functions based on the headers you import). Another reason actually was to have just one unique place where the logic is executed, for testing convenience, but anyway..

**Comment by elect86** [comment]
I never saw unsigned matrices so far, so unless we are proven otherwise I'd say let's merge that and continue

Also, I'm porting directly from the native library

Edit: shall we keep the constructor for building matrices by diagonal? That is, ie `Mat3`
```kotlin
	constructor(x: Number, y: Number) : this(x, y, 1)
	constructor(x: Number, y: Number, z: Number) : this(
		x, 0, 0,
		0, y, 0,
		0, 0, z)
```
Edit2: maybe it's better to avoid any operation on generic and write down every specific implementation, such as:
`fun <N : Number> times(res: Mat2T<N, out Vec2T<N>>, a: Mat2T<N, out Vec2T<N>>, b: Mat2T<N, out Vec2T<N>>)`
I'm afraid that in hot loops this may penalizes us substantially..
And we are generating code already, so..

**Reply by ArcheCraft** [reply]
So for the diagonal constructors: I would keep them for the square matrices (same width and height), as that helps making an identity matrix and other useful things.

For the second edit: We definitely should generate all variants seperately, but that should be just a move of the code into each matrix and an edit to replace the generics with the actual type

**Reply by elect86** [reply]
ok then, I uncommented them back

I'm doing the `Unary arithmetic operators` atm

**Reply by elect86** [reply]
Now `// -- Increment and decrement operators --`

**Reply by elect86** [reply]
`Unary arithmetic operators`

**Reply by elect86** [reply]
`// -- Binary operators --`

**Reply by ArcheCraft** [reply]
I degenerified the times function and moved it to objects in the specific implementation classes
That way you can still use the function for custom matrix implementations, but it should be generally more efficient

**Reply by elect86** [reply]
Nice!
Ps: I wrote some binary function (the multiplication is wrong, I'll fix it)

**Reply by ArcheCraft** [reply]
Also why did you limit the Swizzles to 2 params?

**Reply by ArcheCraft** [reply]
> (the multiplication is wrong, I'll fix it)

Multiplication with matrices or with scalars?

**Reply by elect86** [reply]
> Also why did you limit the Swizzles to 2 params?

I'm trying to figuring out which are the bottleneck when running `allTest`. I'm pretty sure swizzle is somehow responsible

> Multiplication with matrices or with scalars?

With vectors, but maybe I didn't push it yet

**Reply by ArcheCraft** [reply]
Should we add `@ExperimentalUnsignedTypes` to all unsigned declarations or add `-Xopt-in=kotlin.ExperimentalUnsignedTypes` to the compiler options?
That would get rid of all the warnings in the log when comiling

**Reply by ArcheCraft** [reply]
> > Also why did you limit the Swizzles to 2 params?
> 
> I'm trying to figuring out which are the bottleneck when running `allTest`. I'm pretty sure swizzle is somehow responsible

I just checked: The swizzles increase the **compile** time by a lot
It's not running the tests, it's the compiling that takes so long, and it compiles 3 times (native, jvm and metadata)

**Reply by ArcheCraft** [reply]
Also, just an idea... What if we would make the swizzles generic? We would only need 3 (Vec2T, Vec3T, Vec4T) instead of 33 (3 * 11 types) and could still have all swizzles while reducing the compile time
And it's like 1200 swizzle variables per Vec4X

**Reply by elect86** [reply]
> Should we add `@ExperimentalUnsignedTypes` to all unsigned declarations or add `-Xopt-in=kotlin.ExperimentalUnsignedTypes` to the compiler options?

I'm for the less invasive option, that is the latter, you?

> I just checked: The swizzles increase the **compile** time by a lot It's not running the tests, it's the compiling that takes so long, and it compiles 3 times (native, jvm and metadata)

I wonder if we move those inside the corresponding classes..
My idea for having them out was to decrease the clutter when the user wanted to exploit the methods on the vector, but I guess this isn't a valid reason, they will be brought in anyway.. or?

> Also, just an idea... What if we would make the swizzles generic? We would only need 3 (Vec2T, Vec3T, Vec4T) instead of 33 (3 * 11 types) and could still have all swizzles while reducing the compile time And it's like 1200 swizzle variables per Vec4X

It's another option, sure we can try

**Reply by ArcheCraft** [reply]
> > Should we add `@ExperimentalUnsignedTypes` to all unsigned declarations or add `-Xopt-in=kotlin.ExperimentalUnsignedTypes` to the compiler options?
> 
> I'm for the less invasive option, that is the latter, you?

Generally i would also be for the second option, except this is a library and that annotation exists for the sole reason to be propagated through apis and then the end user would need to either `@OptIn` or use the compiler argument. But if we accept unsigned vectors as a core feature of this library i would guess it would be ok to use the compiler argument to silence the warnings

> 
> > I just checked: The swizzles increase the **compile** time by a lot It's not running the tests, it's the compiling that takes so long, and it compiles 3 times (native, jvm and metadata)
> 
> I wonder if we move those inside the corresponding classes.. My idea for having them out was to decrease the clutter when the user wanted to exploit the methods on the vector, but I guess this isn't a valid reason, they will be brought in anyway.. or?
> 

If you don't add an import for them, IDEA will only display extension functions from other packages if you double tap STRG + SPACE

> > Also, just an idea... What if we would make the swizzles generic? We would only need 3 (Vec2T, Vec3T, Vec4T) instead of 33 (3 * 11 types) and could still have all swizzles while reducing the compile time And it's like 1200 swizzle variables per Vec4X
> 
> It's another option, sure we can try

I would try that and if it helps, i'll push a commit

**Reply by ArcheCraft** [reply]
I implemented the generic swizzles
It now takes ~20 sec per compile, from before >3 min per compile (on my PC)
I needed to add a generic "Delegate" Vector implementation and i extended the generic dot-product function to support it as well

Also we could consider making the swizzles only a wrapper around the original vector so updates of the original vector would be reflected in the new vector and vice-versa. Do you think this would be useful or would it be too big of a change for users?

**Reply by ArcheCraft** [reply]
So as i anyways reworked the way matrix operations are generated i also did matrix - vector multiplications

**Reply by elect86** [reply]
> I implemented the generic swizzles It now takes ~20 sec per compile, from before >3 min per compile (on my PC) I needed to add a generic "Delegate" Vector implementation and i extended the generic dot-product function to support it as well
> 
> Also we could consider making the swizzles only a wrapper around the original vector so updates of the original vector would be reflected in the new vector and vice-versa. Do you think this would be useful or would it be too big of a change for users?

Can you give me a short example of what you mean?

**Reply by ArcheCraft** [reply]
Something like:
`class Vec2Delegate<T>(x_: KMutableProperty0<T>, y_: KMutableProperty0<T>) : Vec2T<T> {
    override var x by x_
    override var y by y_
}`

`val <T> Vec2T<T>.XY get() = Vec2Delegate(this::x, this::y)`

**Reply by elect86** [reply]
> Do you think this would be useful 

I honestly don't know. If you feel it, you can go ahead. Otherwise we can postpone this later on when/if we get some more feedback about it and focus on other stuff

I opened an [issue](https://youtrack.jetbrains.com/issue/KT-49940) though for the swizzle compile time

**Reply by ArcheCraft** [reply]
Actually on my PC it did run the tests, it just took like 15 Minutes...

**Reply by elect86** [reply]
Uh, ok, but still unfeasible for development

Btw, I noticed [glm](https://github.com/g-truc/glm/tree/master/glm/ext) does indeed have unsigned matrices (beyond float, double and int), shall we do the same? 
Edit: we could leave out the unsigned given it doesnt extend `Number` for the moment, since `Mat*T` requires that constraint

**Reply by elect86** [reply]
Also, what about we move `width` and `height` as top level properties and avoid passing them every-time again and again?

**Reply by ArcheCraft** [reply]
That makes understanding the code much harder as the code flow is not (that) clear
And i would say easy readable code is more important than typing a little bit less
